test/HLSL/passes/multi_dim_one_dim/gep_addrspacecast_gep.ll

```diff
@@ -101,3 +102,97 @@ entry:
   ret void
 }
+
+; Test that we compute the correct index when there are multiple
+; geps after the addrspacecast. This also exercises the case
+; where one of the outer geps ends in an array which hits
+; an early return in MergeGEP.
+; CHECK-LABEL: @merge_gep_multi_level_end_in_sequential_with_addrspace
+; CHECK:  %0 = mul i32 %idx0, 9
+; CHECK:  %1 = add i32 %idx1, %0
+; CHECK:  %2 = getelementptr [2304 x float], [2304 x float] addrspace(3)* @ArrayOfArray.1dim, i32 0, i32 %1
+; CHECK:  %3 = addrspacecast float addrspace(3)* %2 to float*
+; CHECK:  load float, float* %3
+define void @merge_gep_multi_level_end_in_sequential_with_addrspace(i32 %idx0, i32 %idx1) {
+entry:
+  %gep0 = getelementptr inbounds [256 x [9 x float]], [256 x [9 x float]] addrspace(3)* @ArrayOfArray, i32 0
+  %asc  = addrspacecast [256 x [9 x float]] addrspace(3)* %gep0 to [256 x [9 x float]]*
+  %gep1 = getelementptr inbounds [256 x [9 x float]], [256 x [9 x float]]* %asc, i32 0, i32 %idx0
+  %gep2 = getelementptr inbounds [9 x float], [9 x float]* %gep1, i32 0, i32 %idx1
+  %load = load float, float* %gep2
+  ret void
+}
+
+; Test that we compute the correct index when there are three levels of geps.
+; This also exercises the case where one of the outer geps ends in an
+; array which hits an early return in MergeGEP.
+; CHECK-LABEL: @merge_gep_multi_level_end_in_sequential
+; CHECK:  %0 = mul i32 %idx0, 9
+; CHECK:  %1 = add i32 %idx1, %0
+; CHECK:  %2 = getelementptr [2304 x float], [2304 x float] addrspace(3)* @ArrayOfArray.1dim, i32 0, i32 %1
+; CHECK:  load float, float addrspace(3)* %2
+define void @merge_gep_multi_level_end_in_sequential(i32 %idx0, i32 %idx1) {
+entry:
+  %gep0 = getelementptr inbounds [256 x [9 x float]], [256 x [9 x float]] addrspace(3)* @ArrayOfArray, i32 0
+  %gep1 = getelementptr inbounds [256 x [9 x float]], [256 x [9 x float]] addrspace(3)* %gep0, i32 0, i32 %idx0
+  %gep2 = getelementptr inbounds [9 x float], [9 x float] addrspace(3)* %gep1, i32 0, i32 %idx1
+  %load = load float, float addrspace(3)* %gep2
+  ret void
+}
+
+; Test that we compute the correct index when the global has 3 levels of
+; nested arrays and an addrspacecast.
+; CHECK-LABEL: @merge_gep_multi_level_with_addrspace
+; CHECK:  %0 = mul i32 %idx0, 9
+; CHECK:  %1 = add i32 %idx1, %0
+; CHECK:  %2 = mul i32 %1, 3
+; CHECK:  %3 = add i32 %idx2, %2
+; CHECK:  %4 = getelementptr [6912 x float], [6912 x float] addrspace(3)* @ArrayOfArrayOfArray.1dim, i32 0, i32 %3
+; CHECK:  %5 = addrspacecast float addrspace(3)* %4 to float*
+; CHECK:  load float, float* %5
+define void @merge_gep_multi_level_with_addrspace(i32 %idx0, i32 %idx1, i32 %idx2) {
+entry:
+  %gep0 = getelementptr inbounds [256 x [9 x [3 x float]]], [256 x [9 x [3 x float]]] addrspace(3)* @ArrayOfArrayOfArray, i32 0, i32 %idx0
+  %asc  = addrspacecast [9 x [3 x float]] addrspace(3)* %gep0 to [9 x [3 x float]]*
+  %gep1 = getelementptr inbounds [9 x [3 x float]], [9 x [3 x float]]* %asc, i32 0, i32 %idx1
```

Comment:
What happens if there are two addrspacecast instructions, which may or may not cancel each other out?  I don't know how exactly that might be generated from HLSL, but seems like a legal form that MergeGEP might need to handle.

Reply:
The code currently only looks back through one level of addrspacecast. If there were multiple addrspacecasts between the gep it would not merge the geps.

Reply:
This seems like a useful case to include a test for.  It does seem like it could be handled better as well.  

For instance:
```
This:
  addrspace(3) GEP -> addrspacecast(0) -> GEP -> addrspacecast(3)
could simplify to:
  addrspace(3) mergedGEP
instead of:
  addrspace(3) mergedGEP -> addrspacecast(0) -> addrspacecast(3)

And:
  addrspace(5) GEP -> addrspacecast(0) -> GEP -> addrspacecast(3)
could simplify to:
  addrspace(5) mergedGEP -> addrspacecast(3)
instead of:
  addrspace(5) mergedGEP -> addrspacecast(0) -> addrspacecast(3)

Testing both with and without a GEP at the end as well:
GEP -> addrspacecast -> GEP -> addrspacecast -> GEP
```

Reply:
I'm not sure those tests actually add anything new? The addrspacecast will not be merged with the current code.